File: Growth.py
import numpy as np
from scipy.optimize import minimize
from scipy.stats import norm
import matplotlib.pyplot as plt

# Define the PDF function for the skewed normal distribution
def skew_normal_pdf_with_shift(x, xi, omega, alpha):
    t = (x - xi) / omega
    return (2 / omega) * norm.pdf(t) * norm.cdf(alpha * t)

# Define the objective function (minimise the squared error)
def objective_func(params, x_data, y_data):
    xi, omega, alpha= params
    y_pred = skew_normal_pdf_with_shift(x_data, xi, omega, alpha)
    return np.sum((y_data - y_pred)**2)

# Raw data
x_data = np.array([10, 16, 22, 28, 34, 40])
y_data = np.array([0.30, 0.36, 0.34, 0.38, 0.06, 0.00])

# Initial guess parameters [xi, omega, alpha]
initial_params = [25, 20, -10]

# Parameters fitted using quasi-Newton optimisation methods
result = minimize(objective_func, initial_params, args=(x_data, y_data), method='BFGS')

# Get the parameters of the fit
xi_est, omega_est, alpha_est= result.x
print(f'Estimated parameters: xi = {xi_est:.4f}, omega = {omega_est:.4f}, alpha = {alpha_est:.4f}')

# Generate a fitted curve using the fitted parameters
x_fit = np.linspace(0, 50, 5000)  
y_fit = skew_normal_pdf_with_shift(x_fit, xi_est, omega_est, alpha_est)

# Plotting raw data and fitting curves
plt.plot(x_data, y_data, 'ro', label='Original data')
plt.plot(x_fit, y_fit, 'k-', linewidth=5, label='Fitted skew normal distribution with shift')
plt.legend()
plt.xlabel('X')
plt.ylabel('Y')
plt.title('Skew Normal Distribution with Vertical Shift')
plt.show()

File: Growth.r
# 安装和加载sn包
if(!require(sn)) {
  install.packages("sn")
}
library(sn)

# 数据
x <- c(3.5, 4.5, 5.5, 6.5, 7.5, 8.5, 9.5, 10.5, 11.5)
y <- c(0.0000000, 0.3666667, 0.4336667, 0.4556667, 0.5676667, 0.6413333, 0.7153333, 0.3973333, 0.0000000)

# 定义目标函数 (最小化平方误差)
objective_function <- function(params, x, y) {
  xi <- params[1]
  omega <- params[2]
  alpha <- params[3]
  
  # 使用dsn函数计算偏正态分布的y值
  y_pred <- dsn(x, xi, omega, alpha)
  
  # 返回平方误差
  return(sum((y - y_pred)^2))
}

# 初始猜测参数 [xi, omega, alpha]
initial_params <- c(7, 1, 1)

# 使用optim进行拟合 (quasi-Newton method)
fit <- optim(initial_params, objective_function, x = x, y = y, method = "BFGS")

# 获取拟合的参数
xi_est <- fit$par[1]
omega_est <- fit$par[2]
alpha_est <- fit$par[3]

cat("拟合的参数: \n")
cat("xi =", xi_est, "\n")
cat("omega =", omega_est, "\n")
cat("alpha =", alpha_est, "\n")

# 使用拟合的参数生成拟合曲线
x_fit <- seq(min(x), max(x), length.out = 100)
y_fit <- dsn(x_fit, xi_est, omega_est, alpha_est)

# 绘制结果
plot(x, y, pch = 16, col = "red", xlab = "x", ylab = "y", main = "Skew Normal Distribution Fit")
lines(x_fit, y_fit, col = "blue", lwd = 2)
legend("topright", legend = c("Observed data", "Fitted curve"), col = c("red", "blue"), pch = c(16, NA), lty = c(NA, 1))